sherpa/models/tests/test_model_op.py

```diff
@@ -695,3 +700,268 @@ def test_explicit_numpy_combination():
+                          (((GAUSS1 + GAUSS2 - 4) / (BOX1 + BOX2) * SCALE1),
+                           ["g1 / (b1 + b2) * s1",
+                            "g2 / (b1 + b2) * s1",
+                            "-(4.0) / (b1 + b2) * s1"]),
+                           ((BOX1 + BOX2) ** 2 / 2 + GAUSS1,
+                            ["(b1 + b2) ** 2.0 / 2.0",
+                             "g1"]),
+                           (GAUSS1 + (BOX1 - BOX2) ** 2 / 2,
+                            ["g1",
+                             "(b1 - b2) ** 2.0 / 2.0"])
+                          ])
+def test_model_deconstruct(model, expecteds):
+    """Check model deconstruction using the model name and evaluation.
+
+    This uses a set of known models in the assumption we have covered
+    all the relevant code paths. Particular care is needed to ensure
+    conditions like m1 / (m2 + m3) are included. At some point this
+    becomes a regression test, as it's not neessarily important that
+    the best possible" deconstruction is created, just that we get
+    consistent results (e.g. given that the code does not simplify
+    expressions, in particular the handling of negation).
+
+    The evaluation test checks that
+
+        model(x) = sum_i term_i(x)
+
+    for all the terms that model_deconstruct creates. The idea is that
+    the models are set to non-zero values over the range use (ie
+    within -10 to 10 for the box components), and that there are some
+    components which vary with x just to check everything passes
+    through correctly.
+
+    """
+
+    terms = model_deconstruct(model)
+    assert len(terms) == len(expecteds)
+    for term, expected in zip(terms, expecteds):
+        assert term.name == expected
+
+    x = np.arange(-9, 9, 1)
+    got_model = model(x)
+    got_terms = np.zeros_like(got_model)
+    for term in terms:
+        got_terms += term(x)
+
+    assert got_terms == pytest.approx(got_model)
+
+
+def test_model_deconstruct_possible_recursion_error_lhs():
+    """Try and test the recursion-handling on the LHS.
+
+    It is not obvious if the recursion-handling is always going to
+    trigger when 1000 frames are hit, so we try a value which is
+    known - for Python 3.11 - to trigger a recursion error, but it
+    may not with other Python cases.
+
+    """
+
+    def mk(n):
+        return basic.Scale1D(f"c{n}")
+
+    mdl = mk(0)
+    for i in range(1, 1001):
+        mdl += mk(i)
+
+    # mdl is (((...(c0 + c1) + .. ) + c999) + c1000)
+    #
+    cpts = model_deconstruct(mdl)
+
+    # There are 1001 individual components, so this should return each
+    # of them, unless we hit the recursion-error check, at which point
+    # the deconstruction will stop.
+    #
+    ncpts = len(cpts)
+    if ncpts == 1001:
+        # No recursion error found. Is there a nice way to identify
+        # this so we can review? For now, just treat is as a pass.
```

Comment:
I disagree. If no recursion error has been found then I don't think we want the code to start failing (I'm thinking of a time when sherpa is no longer being developed but it can still be built with the latest python).

I'm going to rework the test to use 101 and 1001 components so we at least explicitly check (for python 3.11-era) both the "no recursion" and "yay recursion" cases, with a little bit of commentary to see if that helps.